Add cached dependency manager

mvdbeek opened the pull request:
Similar to https://github.com/galaxyproject/galaxy/pull/2986, implements a mechanism that allows tool dependencies to be cached.

If the `use_cached_dependency_manager` option is set to True in `galaxy.ini`, we build a hash of the combination of a tools' requirements, and store the resulting environment in a directory specified by the
`tool_dependency_cache_dir` option in `galaxy.ini`.

The approach here is only caching (combinations of) conda dependencies when installing new tools through the ToolShed.
When a dependency combination is not yet cached, `conda create` will be invoked with `tool_dependency_cache_dir/<hash>` folder, and otherwise  `conda install` with `tool_dependency_cache_dir/<hash>`.

If a `__name@version` dependency exists, but not a cached environment at job runtime, that environment will be used. So to benefit from cached environments for already installed tools, those cached environment need to be created. Currently this can only be done by (re-)installing tools.

ping @abretaud @jmchilton

mvdbeek committed:
Add cached dependency manager

Similar to https://github.com/galaxyproject/galaxy/pull/2986, implement
a mechanism that allows tool dependencies to be cached.

If the `use_cached_dependency_manager` option is set to True in
galaxy.ini, we build a hash of the combination of a tools' requirements,
and store the resulting environment in a directory specified by the
`tool_dependency_cache_dir` option in galaxy.ini.

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -213,6 +213,19 @@ paste.app_factory = galaxy.web.buildapp:app_factory
 # if it cannot find a local copy and conda_exec is not configured.
 #conda_auto_init = False
 
+# Certain dependency resolvers (namely conda) take a considerable amount of
```
s/conda/Conda/

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -213,6 +213,19 @@ paste.app_factory = galaxy.web.buildapp:app_factory
 
+# Certain dependency resolvers (namely conda) take a considerable amount of
+# time to build an isolated job environment in the job_working_directory.  Set
+# the following option to true to cache the dependencies in a folder. This
```
s/true/True/

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -213,6 +213,19 @@ paste.app_factory = galaxy.web.buildapp:app_factory
+# the following option to true to cache the dependencies in a folder. This
+# option is beta and should only be used if you experience long waiting times
+# before a job is actually submitted to your cluster. If you activate this
+# option, and you install new dependencies you may need to clear out old cached
```
Move comma after `dependencies`.

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -213,6 +213,19 @@ paste.app_factory = galaxy.web.buildapp:app_factory
+
+# By default the tool_dependency_cache_dir is the _cache directory
+# of the tool dependency directory
+#tool_dependency_cache_dir = tool_dependency_dir/_cache
```
`s/tool_dependency_dir/<tool_dependency_dir>/`

mvdbeek replied:
hmm, so

```
tool_dependency_cache_dir = _cache
```

?
I would like to indicate that by default we store the cache in tool_dependency_dir/_cache.
Maybe 

```
tool_dependency_cache_dir = <tool_dependency_dir>/_conda
```

would be better?

nsoranzo replied:
I guess you saw my comment before I edited it to add quotes, GitHub hides everything between the angle brackets.
Anyway, I suggested to use

`<tool_dependency_dir>/_cache`

mvdbeek replied:
sorry, forgot about that GH quirk :)

mvdbeek committed:
Better phrasing of cache options in galaxy.ini.sample (Thanks @nsoranzo)

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -213,6 +213,19 @@ paste.app_factory = galaxy.web.buildapp:app_factory
 
+# Certain dependency resolvers (namely Conda) take a considerable amount of
+# time to build an isolated job environment in the job_working_directory if the
+# job working diretory is on a network share.  Set the following option to True
```
s/diretory/directory/

mvdbeek committed:
Rework cached dependencies to be more robust

to updates or changes in depedencies, folder structure and resolver
configuration. Instead of hashing name, type and version of a
dependency, hash the json representation of the dependencies returned by
the dependency resolver, which include the path to the environment and
the depedency type. This is only applied to resolvers whose cacheable
attribute is set to True (conda-only, currently).

nsoranzo commented on lib/galaxy/tools/deps/__init__.py:
```diff
@@ -155,3 +165,55 @@ def __parse_resolver_conf_xml(self, plugin_source):
+                # while other resolvers may not create the hashed_requirements_dir
+                os.mkdir(hashed_requirements_dir)
+            with open(os.path.join(hashed_requirements_dir, 'dep_commands.sh'), 'w') as cmds_f:
+                [cmds_f.write("%s\n" % line) for line in commands]
```
This looks "perlish":

``` python
for line in commands:
    cmds_f.write("%s\n" % line)
```

bgruening commented on config/galaxy.ini.sample:
```diff
@@ -213,6 +213,19 @@ paste.app_factory = galaxy.web.buildapp:app_factory
+# used if you experience long waiting times before a job is actually submitted
+# to your cluster. If you activate this option and install or remove dependencies,
+# you may need to clear out old cached environments
+#use_cached_dependency_manager = False
```
Should we name this something like _conda_?
People could be confused if they are using Docker.

mvdbeek replied:
In principle this works for all non-container resolver types, except that you get the biggest benefits with conda. But maybe I should find another angle and do the caching only for conda ... I kind of like inheriting from DependencyManager though.
Let's see if I can come up with an elegant way to do this for conda only. That would reduce the number of environments that would be cached.

mvdbeek commented:
In my testing this is working pretty well:
If I uninstall a dependency, it is not being used from the cache (because the resolver doesn't find it), if I reinstall it it is being picked up from the cache, if I downgrade a dependency inside the `__<name>@<version>` environment, the downgraded version will be used (conda install into the pre-existing hash dir) etc. So whatever happens, we do not diverge from the packages that would be chosen by the dependency resolver.

jmchilton commented:
Sorry I didn't respond sooner - I was at a conference last week.

I appreciate the effort and I know this is solving a real problem, it isn't exactly how I would have gone about it - kind of close though.

For instance this doesn't handle locking well right? There could be a race condition there because the bash stuff is pretty basic? I'd also hope the cache conda dependencies thing would sit beside conda somewhere by default to solve @natefoo's linking problems (this PR does that I think) and would cause the cached environment creations during TS installs (to solve @natefoo's RO problems).

What do you think about this modification to the approach? The TS piece doesn't need to be apart of this - but I'm a bit worried about the race condition.

mvdbeek commented:
I think conda does have a mechanism to prevent race conditions -- I have
seen lock file creation when creating / updating the environment, but I can
check how robust this is when I fire a lot of parallel jobs before the
environment is cached.

I can definitely add the cache creation during TS install, that should be
just calling the build_dependency_shell commands.

On Oct 31, 2016 2:19 PM, "John Chilton" notifications@github.com wrote:

> Sorry I didn't respond sooner - I was at a conference last week.
> 
> I appreciate the effort and I know this is solving a real problem, it
> isn't exactly how I would have gone about it - kind of close though.
> 
> For instance this doesn't handle locking well right? There could be a race
> condition there because the bash stuff is pretty basic? I'd also hope the
> cache conda dependencies thing would sit beside conda somewhere by default
> to solve @natefoo https://github.com/natefoo's linking problems (this
> PR does that I think) and would cause the cached environment creations
> during TS installs (to solve @natefoo https://github.com/natefoo's RO
> problems).
> 
> What do you think about this modification to the approach? The TS piece
> doesn't need to be apart of this - but I'm a bit worried about the race
> condition.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/galaxyproject/galaxy/pull/3106#issuecomment-257291195,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AGfVpQ71kxXe_7EE33DeBW77kavevDTrks5q5erSgaJpZM4KjhXz
> .

mvdbeek commented:
Hmm, conda is actually locking, both for the creation of the environment as well as when calling conda install on existing environments. that means only 1 job at a time is being prepared for submission :(.
I guess creating the cached environments at tool install time is the way to go, and then skipping the environment building when creating jobs. Which means that manual changes to the `__name@version` environments will not be automatically propagated to the cache. I guess that would be good enough.

nsoranzo commented:
> Which means that manual changes to the `__name@version` environments will not be automatically propagated to the cache. I guess that would be good enough.

Do we actually still need the `__name@version` environments?

mvdbeek commented:
> Do we actually still need the __name@version environments?

Good question, I'm not sure. Maybe not; they live within the `conda_prefix/envs` dir and 
- _may_ be necessary just so that there won't be any conflicts inside the root environment.
- we _may_ need them because of the `--offline` environment creation.
- if we keep them, people can just disable the cached environments without needing to re-install conda dependencies

Do you know more about this, @bgruening ? 
Maybe we can directly create these cached environments. But even if we eliminate the `__name@version`environments, it will be hard to do any manual maintenance work. Perhaps we can add functionality to planemo or ephemeris to do this?!

mvdbeek commented:
Hmm, this is getting more and more complicated: If we cache the environments at install time we need to get all combinations of requirements for all tools in a repository ... it's too bad we can't just sequentially chain those `__name@version` environments ... or can we?

nsoranzo commented:
> Good question, I'm not sure. Maybe not; they live within the `conda_prefix/envs` dir and 
> - _may_ be necessary just so that there won't be any conflicts inside the root environment.

Not sure I understand, which root environment? All packages are installed in `conda_prefix/pkgs/`, possibly in multiple versions, conflicts should matter only when creating the env for job execution.

> - we _may_ need them because of the `--offline` environment creation.

I don't think that is the case, I removed all envs in my conda install and tried:

```
~/miniconda2/bin/conda create --unknown --offline -n __pandoc pandoc==1.17.2
```

and it worked flawlessly. Pandoc pkg was already installed on my system as a dependency for Prokka.

> - if we keep them, people can just disable the cached environments without needing to re-install conda dependencies

I think `__name@version` envs were needed only for reproducibility (i.e. to make sure the dependencies of the tool requirement are not upgraded at job creation), which would be substituted by your caching mechanism. So I think this shouldn't probably be an option.

> If we cache the environments at install time we need to get all combinations of requirements for all tools in a repository

We need a cached environment per installed tool at the most (multiple tools can have the same set of requirements). Obviously a tool can be installed at multiple versions, but I don't think that this will be much bigger than the current set of envs.

mvdbeek commented:
> We need a cached environment per installed tool at the most (multiple tools can have the same set of requirements).

That's right, but there isn't an obvious place where we would have the tools in the repo represent their individual requirement combinations.
A repository may declare a dependency on bedtools and samtools, but individual tools may need one or the other or both, and each of these cases must be cached.

> I think __name@version envs were needed only for reproducibility (i.e. to make sure the dependencies of the tool requirement are not upgraded at job creation), which would be substituted by your caching mechanism. So I think this shouldn't probably be an option.

Right, but how do we get the packages in the first place? We currently get them as a side-effect of creating the `__name@version` env in https://github.com/galaxyproject/galaxy/blob/bf05ab375dd60c46787593ff7c87b21db9e0f056/lib/galaxy/tools/deps/conda_util.py#L351

mvdbeek commented:
> Right, but how do we get the packages in the first place? We currently get them as a side-effect of creating the __name@version env in https://github.com/galaxyproject/galaxy/blob/bf05ab375dd60c46787593ff7c87b21db9e0f056/lib/galaxy/tools/deps/conda_util.py#L351

OK, it would probably work if we directly create the cached environment.

mvdbeek committed:
Build dependency cache at install time

and only activate cached environments if they exist.

mvdbeek commented:
@nsoranzo @jmchilton 
The current version still creates the `__name@version` environments, and then creates for each tool a cached environment (if it doesn't exist yet) during the installation (The caching would be easily exposable through the API, if that's considered useful). When running a tool, if a cached environment exists, it will be used, otherwise we fall back to creating a `conda-env` and `conda-metadata-env` in the job_working_directory (So no cache creation when running a tool, and no race-condition/concurrency problems). Does that seem like a good compromise?

Otherwise I would need to first run through the installable resolvers to see which dependency _would_ be installed using a cacheable dependency resolver (because we need the list of dependencies that should be resolved using conda, which is possibly not equivalent to the full dependency list), and then trigger the installation using the calculated hash.
Let me know if you see another path that we could follow.

mvdbeek committed:
Only attempt to build cache once per install

and override __eq__ for ToolRequirement, to simplify
checking if ToolRequirements are already installed/cached.

lecorguille commented:
👍 NEED

jmchilton commented:
@mvdbeek This seems okay I think. Can you rebase or merge to resolve conflicts? Is there any reasons this shouldn't be enabled by default?

mvdbeek commented:
> Is there any reasons this shouldn't be enabled by default?

It's a bit of a departure from what we have described in the admin documentation. I'd say let's get this in as an option for those admins that already know that they need it, and then make planemo create those environments (`planemo conda_install` would need to be updated, I think), add an API to create, update and delete environments and finally the UI, and somewhere along those lines we can make it default.

Btw, if we manage to eliminate the need for the `__name@version` environment, I think we'll also get rid of the need for the offline mode, so we can finally update conda.

mvdbeek commented:
@jmchilton I've rebased the PR and updated the PR description.

natefoo commented:
I believe this is exactly what I need to run these directly from CVMFS. I can give it a test if I can remember one of the tools that had the problems.

mvdbeek commented:
Conflicting again? grrrr

mvdbeek commented:
OK, rebased it again.

mvdbeek committed:
Fix NameError

jmchilton commented:
Thanks for the PR - this was really needed!

lparsons commented:
So, most of my conda installed tools are failing with the error `Conda dependency seemingly installed but failed to build job environment.` which means I'm desperate for a fix. Is there a recommended way to install this beyond just moving to a bleeding edge build?

lparsons commented:
Perhaps we could get this into `release_16.10`? Conda dependencies simply do not work (reliably) without it.

mvdbeek commented:
It's working well for me, but for convenience we should also backport https://github.com/galaxyproject/galaxy/pull/3222. Otherwise you would need to re-install these tools.

lparsons commented:
Hmmm, what are you doing to avoid the `... failed to build job environment` error? Seems to happen whenever I run jobs that use a conda dependency over a collection. At this point, I'm thinking I'll have to deactivate all tools that use conda deps and either hold off on them 😦  or perhaps write module files or something to avoid using conda. Any tips would be greatly appreciated. 

BTW, thanks so much for this PR! Can't wait to be able to use conda!

lparsons commented:
Also, I didn't think `release_16.10` was out yet, so not really backporting... 😉

nsoranzo commented:
@lparsons 16.10 is not out, but it's feature freezed. Whether this is a new feature or a bug fix, or if an exception should be granted, is up for discussion.

lparsons commented:
@nsoranzo Well, since it wouldn't be enabled by default, I think it's relatively safe. And it's certainly attempting to resolve a showstopper of a bug for me, so in some sense it is a "bug fix". ;-)

mvdbeek commented:
> Hmmm, what are you doing to avoid the ... failed to build job environment error? Seems to happen whenever I run jobs that use a conda dependency over a collection

Sorry, I wasn't clear, I meant that I'm using this in production, so I would say this is (probably) not going to break stuff all over the place. I opened PR #3227 for the backport.

lparsons commented:
@mvdbeek You rock! Thanks.
